pkg/netpol/connlist/exposure_analysis.go

```diff
@@ -0,0 +1,159 @@
+	res := make([]XgressExposureData, len(xgressExp))
+	for i := range xgressExp {
+		res[i] = xgressExp[i]
+	}
+	return res
+}
+
+func (ep *exposedPeer) IsProtectedByIngressNetpols() bool {
+	return ep.pExpData.isIngressProtected
+}
+
+func (ep *exposedPeer) IngressExposure() []XgressExposureData {
+	return xgressExposureListToXgressExposureDataList(ep.pExpData.ingressExposure)
+}
+
+func (ep *exposedPeer) IsProtectedByEgressNetpols() bool {
+	return ep.pExpData.isEgressProtected
+}
+
+func (ep *exposedPeer) EgressExposure() []XgressExposureData {
+	return xgressExposureListToXgressExposureDataList(ep.pExpData.egressExposure)
+}
+
+// ----------------------------------------------------
+
+// gets an exposure map and builds ExposedPeer list while refining exposure connections
+// for each peer refines its exposure data as followed:
+// * if a peer is not protected by netpols on any direction (ingress/egress) :
+// the xgress lists are already empty so will have an empty xgressExp list (output will display not protected)
+// * if a peer is protected and exposed to entire cluster on all conns:
+// the list will include only this (eliminates other conns with specific namespaces)
+// * if a peer is exposed to entire cluster on ingress/egress on a specific connection:
+// eliminate exposure-data to specific potential namespaces on same direction and same connection
+func buildExposedPeerListFromExposureMap(exposuresMap exposureMap) []ExposedPeer {
+	res := make([]ExposedPeer, 0)
+	for p, expData := range exposuresMap {
+		ingExp := make([]*xgressExposure, 0)
+		egExp := make([]*xgressExposure, 0)
+		if expData.isIngressProtected {
+			ingExp = append(ingExp, loopAndRefineXgressData(expData.ingressExposure)...)
+		}
+		if expData.isEgressProtected {
+			egExp = append(egExp, loopAndRefineXgressData(expData.egressExposure)...)
+		}
+		// final peer's exposure data
+		expInfo := &exposedPeer{
+			peer: p,
+			pExpData: &peerExposureData{
+				isIngressProtected: expData.isIngressProtected,
+				isEgressProtected:  expData.isEgressProtected,
+				ingressExposure:    ingExp,
+				egressExposure:     egExp,
+			},
+		}
+		res = append(res, expInfo)
+	}
+	return res
+}
+
+func loopAndRefineXgressData(xgressData []*xgressExposure) []*xgressExposure {
+	res := make([]*xgressExposure, 0)
+	// helping var
+	var entireClusterConn conn.Connection
+	for _, singleConn := range xgressData {
+		//  exposed to entire cluster on all conns - result will include this one general exposureConn
+		if singleConn.exposedToEntireCluster && singleConn.potentialConn.AllConnections() {
+			res = nil // remove previous conns if exist
+			res = append(res, singleConn)
+			break
+		}
+		if singleConn.exposedToEntireCluster {
+			entireClusterConn = singleConn.potentialConn
+			// refine result - exclude data to/from specific ns with same conn value
+			res = refineListConnsContainedInEntireConn(res, entireClusterConn)
```

Comment:
this refinement can be done at the end in one iteration only

Reply:
done